_posts/2016-05-25-coordination-in-akka.md

```diff
@@ -0,0 +1,155 @@
+
+<img src="/assets/img/custom/blog/coordination-diagram.png" alt="Coordination" title="Coordination" class="img img-center img-responsive style-screengrab">
+
+The FSM needs a reference into the coordinator, as it's responsible for managing the lifecycle of the FSMs. Coordinator sends a message to the FSM telling to flush its internal state. At the same time puts back the FSM into the pool.
+
+## Creating actors
+
+Remember how the FSM notifies that its work has finished:
+
+```scala
+private def finishWorkWith(message: Any) = {
+   replyTo ! message
+   coordinator ! message
+   goto(Idle)
+}
+```
+
+`replyTo` is a variable that was initialised with `sender()` when receiving the first `ItemReported` message. Remember that the message was forwarded so the sender is the Route Actor. `coordinator` is an ActorRef that has been injected when the actor was created. We're using poor man's dependency injection for that purpose. Let's see the definition of our actor factory:
+
+```scala
+lazy val itemFSMFactory: (ActorContext, ActorRef) => ActorRef =
+ (context, self) => {
+   val itemFSM = ItemFSM.props(itemReportedProducer, itemDeletedBus, coordinator = self)
+   context.actorOf(itemFSM)
+ }
+```
+
+This factory will be called n times (depending on the size of the pool) when creating the Actor Pool.
+
+## Supervising actors
+
+Our coordinator is in charge of the pool and what to do when the FSM finishes in an expected way. However we didn't talk about what happens when our actor fails. Akka has a built-in mechanism for dealing with failures called [supervision](http://doc.akka.io/docs/akka/current/general/supervision.html). There are different recovery strategy, we'll go for the `Restart` strategy:
+
+```scala
+override val supervisorStrategy =
+  OneForOneStrategy(maxNrOfRetries = 3, withinTimeRange = 15 seconds) {
+    case e: Exception =>
+      actorPool.putBack(sender())
+      SupervisorStrategy.Restart
+  }
+```
+
+As the `Restart` documentation says:
+
+> Discards the old Actor instance and replaces it with a new one, then resumes message processing.
+
+This strategy has been defined in the coordinator. How can we specify that this coordinator is the parent of the FSM to be supervised? Remember how we created the FSM:
+
+```scala
+context.actorOf(itemFSM)
+```
+
+Let's see `actorOf` Scaladoc:
+
+> Create new actor as the child actor of this context
+
+`itemFSMFactory` is called in the scope of the coordinator. Every single actor has a context that includes the sender, self and different meta objects. Creating an actor using coordinator's context bind those actors as parent-child.
+
+## Summary
+
+Thanks to Akka constructs and patterns, and to Scala functional and async goodies we could write coordination code quite easily. In the next post we'll se how to test Akka.
```

Comment:
to Scala functional and async goodies ==>  the functional and async goodies of Scala